hail/src/main/scala/is/hail/io/tabix/TabixReader.scala

```diff
@@ -0,0 +1,361 @@
+package is.hail.io.tabix
+
+import is.hail.HailContext
+import is.hail.io.compress.BGzipInputStream
+import is.hail.utils._
+
+import htsjdk.tribble.util.{ParsingUtils, TabixUtils}
+import org.apache.{hadoop => hd}
+import org.apache.hadoop.io.compress.SplitCompressionInputStream
+import org.apache.spark.SparkContext
+
+import java.io.InputStream
+import java.nio.{ByteBuffer, ByteOrder}
+import java.nio.charset.StandardCharsets
+import java.util.Arrays
+
+import scala.collection.mutable.HashMap
+import scala.language.implicitConversions
+
+// Helper data classes
+
+case class Tabix(
+  val format: Int,
+  val colSeq: Int,
+  val colBeg: Int,
+  val meta: Int,
+  val seqs: Array[String],
+  val chr2tid: HashMap[String, Int],
+  val indices: Array[(HashMap[Int, Array[TbiPair]], Array[Long])]
+)
+
+case class TbiPair(var _1: Long, var _2: Long) extends java.lang.Comparable[TbiPair] {
+  @Override
+  def compareTo(other: TbiPair) = TbiOrd.compare(this, other)
+}
+
+object TbiPair {
+  implicit def tup2Pair(t: (Long, Long)): TbiPair = new TbiPair(t._1, t._2)
+}
+
+object TbiOrd extends Ordering[TbiPair] {
+  def compare(u: TbiPair, v: TbiPair) = if (u._1 == v._1) {
+    0
+  } else if (less64(u._1, v._1)) {
+    -1
+  } else {
+    1
+  }
+
+  def less64(u: Long, v: Long) = (u < v) ^ (u < 0) ^ (v < 0)
+}
+
+// Tabix reader
+// Tabix file format is described here: https://samtools.github.io/hts-specs/tabix.pdf
+
+object TabixReader {
+  val MaxBin: Int = 37450
+  val TadLidxShift: Int = 14
+  val DefaultBufferSize: Int = 1000
+  val Magic: Array[Byte] = Array(84, 66, 73, 1) // "TBI\1"
+
+  def readInt(is: InputStream): Int =
+    (is.read() & 0xff) |
+    ((is.read() & 0xff) << 8) |
+    ((is.read() & 0xff) << 16) |
+    ((is.read() & 0xff) << 24)
+
+  def readLong(is: InputStream): Long =
+    (is.read() & 0xff).asInstanceOf[Long] |
+    ((is.read() & 0xff).asInstanceOf[Long] << 8) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 16) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 24) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 32) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 40) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 48) |
+    ((is.read() & 0xff).asInstanceOf[Long] << 56)
+
+  def readLine(is: InputStream): String = readLine(is, DefaultBufferSize)
+
+  def readLine(is: InputStream, initialBufferSize: Int): String = {
+    val buf = new ArrayBuilder[Byte](initialBufferSize)
+    var c = is.read()
+    while (c >= 0 && c != '\n') {
+      if (c == '\r') {
+        c = is.read()
+        if (c != '\n') fatal("invalid '\\r' character in line")
```

Comment:
Hmm.  I would have thought \r without a following \n should just stay in the line.

Reply:
From the vcf spec (4.3 which is the only one to say anything about it):
 Line separators must be CR+LF or LF and they are allowed only as line separators at end of line.
I take this to mean CR with no LF is an error.

Reply:
I'm not sure what "they are allowed only as line separators at end of line" means, it sounds like nonsense.  "they" clearly refers to CR+LF and CR, so I don't see how this says anything about LF.

The famed Hadoop line splitting code, for example, leaves in LF, since it is not a line separator.  I think we should keep doing the same.

Reply:
I think this should be good now. I should not be the one to dismiss the review, since that will just approve this PR.